### PR TITLE
[#75] Run MicroPython tests in CI

### PR DESCRIPTION
**Pull request by krzema12**
Adding data about MicroPython to the SVG git plot will be done in another PR. Here we deal only with text reports.

### CODE REVIEW

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -9,13 +9,20 @@ on:
...
+          - microPythonTest
     steps:
       - name: Check out
         uses: actions/checkout@v2
         with:
           fetch-depth: 0 # Fetch the whole git history, to be able to calculate some stats based on it.
       - name: Install Kotlin for scripting
         run: sudo snap install --classic kotlin
+      - name: Install MicroPython
+        run: sudo snap install micropython
```
Comment (krzema12) [inline .github/workflows/build_and_test.yml:25]: Ideally it should happen only for MicroPython, but in practice it takes only a handful of seconds so I consider it a fair simplification of the CI script.

Comment (SerVB) [reply .github/workflows/build_and_test.yml:25]: What about global warming? Looked like we could easily add `if` here 😅

Well, feel free to leave it as is actually. I'm just a perfectionist.

File: python/experiments/generate-box-tests-reports.main.kts [hunk 2]
```diff
@@ -26,19 +26,24 @@ import javax.xml.parsers.DocumentBuilderFactory
...
 fun pathFromNamedArgument(argumentName: String): Path? =
+    stringFromNamedArgument(argumentName)?.let { Paths.get(it) }
+
+fun stringFromNamedArgument(argumentName: String): String? =
     args.firstOrNull { it.startsWith("--$argumentName=") }
-        ?.let { Paths.get(it.split("=")[1]) }
+        ?.let { it.split("=")[1] }
 
+val testTask: String = stringFromNamedArgument("test-task")
+    ?: throw IllegalArgumentException("'test-task' parameter is required!")
```
Comment (SerVB) [inline python/experiments/generate-box-tests-reports.main.kts]: Could we leave default value as "pythonTest" here? Would be handy. We can add a warning that it's used when it's not specified.

```suggestion
val testTask: String = stringFromNamedArgument("test-task")
    ?: "pythonTest".also { println("'test-task' parameter is missing, using default 'pythonTest'") }
```

**Review by SerVB — APPROVED** [review]
Only two minor suggestions, please feel free to decide on your own and merge.

By the way, it's getting more difficult to update the both targets. First of all, our main target remains plain Python, right? And MicroPython is just an indicator of current state?

We could, for example, move reports generation to CI, to a special branch for it. I've done it for GH Pages [here](https://github.com/JetBrains/projector-client/blob/79737a7fd834d4e8fdef636eb9224dbe138fef57/.github/workflows/gh-pages.yml#L40-L46), should be similar in our case. If you like the idea, I can try doing it.

File: .github/workflows/build_and_test.yml (hunk 1 above)
Comment (krzema12) [reply .github/workflows/build_and_test.yml:25]: Time spent on looking how to add an `if` to this YAML-based DSL exceeds the cost of leaving it here :P

**Comment by krzema12** [comment]
> By the way, it's getting more difficult to update the both targets. First of all, our main target remains plain Python, right? And MicroPython is just an indicator of current state?
> 
> We could, for example, move reports generation to CI, to a special branch for it. I've done it for GH Pages [here](https://github.com/JetBrains/projector-client/blob/79737a7fd834d4e8fdef636eb9224dbe138fef57/.github/workflows/gh-pages.yml#L40-L46), should be similar in our case. If you like the idea, I can try doing it.

Yeah, I realize that MicroPython means some extra complexity and burden when contributing :sweat: I also thought about automating report generation, like you propose. I propose to merge it as is for now, and we can have a follow-up story to move it to CI. It would be cool if you could give it a try.

Main target - yes, Python. However, I hope that we'll ultimately also support MicroPython. We should actually already start taking it into account when making changes in the compiler.

**Pull request merged**